File: cmd/enrich.ts
// `f5kb enrich` — fills article bodies for the document types the search index
// leaves empty. Behavior reference: enrich_bodies.ts main(). The driver writes
// _enrich_report.json and prints the per-type DONE summaries / GitHub-auth note.
// Returns 0 normally.
//
// Flags:
//   --dump=DIR        dump directory (default outputs/dump)
//   --types="A,B"     subset of implemented type keys
//   --concurrency=N   parallel fetches (default 4)
//   --delay-ms=N      min delay per worker between requests (default 200)
//   --limit=N         cap articles per type (testing)
//   --refetch         re-fetch even if a body / error already present
//   --refetch-errors  re-process only previously-errored articles

import { type ParsedArgs } from "../lib/args.ts";
import { flagBool, flagNum, flagStr } from "../lib/args.ts";
import { type Logger } from "../lib/logger.ts";
import { HttpClient } from "../lib/http/fetcher.ts";
import { enrichDump } from "../lib/enrich/driver.ts";
import { Changelog, changelogPathFromFlag } from "../lib/changelog.ts";

export async function run(args: ParsedArgs, logger: Logger): Promise<number> {
  const flags = args.flags;

  const dump = flagStr(flags, "dump", "outputs/dump")!;
  const types = typeof flags.types === "string"
    ? flags.types.split(",").map((s) => s.trim()).filter(Boolean)
    : null;
  const concurrency = Math.max(1, flagNum(flags, "concurrency", 4) || 4);
  const delayMs = Math.max(0, flagNum(flags, "delay-ms", 200) ?? 200);
  const limit = flags.limit ? (parseInt(String(flags.limit), 10) || null) : null;
  const refetch = flagBool(flags, "refetch");
  const refetchErrors = flagBool(flags, "refetch-errors");

  let githubToken: string | undefined;
  try {
    githubToken = Deno.env.get("GITHUB_TOKEN") || undefined;
  } catch {
    // --allow-env not granted; GitHub enrichment runs unauthenticated.
    githubToken = undefined;
  }

  const http = new HttpClient({ logger: logger.child("http") });

  const changelogPath = changelogPathFromFlag(flags["changelog"], dump);
  const changelog = new Changelog(changelogPath, new Date().toISOString());

  await enrichDump({
    dump,
    types,
    concurrency,
    delayMs,
    limit,
    refetch,
    refetchErrors,
    http,
    githubToken,
    logger,
    changelog,
  });

  await changelog.flush();
  if (changelogPath) logger.info(`Changelog: ${changelogPath}`);

  return 0;
}

File: python/f5kb/cmd/enrich.py
"""f5kb enrich subcommand."""

from __future__ import annotations

import os
import sys

import click

from f5kb.lib.changelog import Changelog, changelog_path_from_flag
from f5kb.http.fetcher import HttpClient
from f5kb.enrich.driver import enrich_dump


@click.command()
@click.option("--dump", "dump_dir", default="outputs/dump", show_default=True)
@click.option("--types", default=None, help="Comma-separated type keys.")
@click.option("--exclude-types", default=None)
@click.option("--concurrency", type=int, default=4, show_default=True)
@click.option("--delay-ms", type=int, default=200, show_default=True)
@click.option("--limit", type=int, default=0, help="Cap per type (0=all).")
@click.option("--refetch", is_flag=True)
@click.option("--refetch-errors", is_flag=True)
@click.option("--changelog", "changelog_flag", default=None, is_flag=False, flag_value="")
@click.option("--yes", is_flag=True, help="Bypass gate: overwrite existing bodies.")
@click.pass_context
def enrich_cmd(ctx, dump_dir, types, exclude_types, concurrency, delay_ms, limit,
               refetch, refetch_errors, changelog_flag, yes):
    """Fetch article bodies for types the search index leaves empty."""
    log = ctx.obj["logger"]
    include_types = [t.strip() for t in types.split(",")] if types else None
    excl_types = [t.strip() for t in exclude_types.split(",")] if exclude_types else None
    github_token = os.environ.get("GITHUB_TOKEN")
    http = HttpClient(logger=log.child("http"))
    cl_path = changelog_path_from_flag(changelog_flag, dump_dir)
    import datetime
    run_id = datetime.datetime.now(datetime.timezone.utc).isoformat().replace("+00:00", "Z")
    changelog = Changelog(cl_path, run_id)

    reports = enrich_dump(
        dump_dir,
        types=include_types,
        exclude_types=excl_types,
        http=http,
        github_token=github_token,
        concurrency=max(1, concurrency),
        delay_ms=max(0, delay_ms),
        limit=limit or None,
        refetch=refetch,
        refetch_errors=refetch_errors,
        logger=log,
        changelog=changelog,
        approval=not yes,
    )
    changelog.flush()
    if cl_path:
        log.info(f"Changelog: {cl_path}")
    total_staged = sum(r.staged for r in reports)
    if total_staged:
        log.warn(f"{total_staged} re-fetched body(ies) STAGED to {dump_dir}/_pending/")
